node/narwhal/tests/common/primary.rs

```diff
@@ -12,23 +12,193 @@
+    }
+}
+
+impl DerefMut for TestPrimary {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.primary
+    }
+}
+
+impl TestPrimary {
+    pub fn fire_cannons(&mut self) {
+        let solution_handle = fire_unconfirmed_solutions(self.sender.as_mut().unwrap(), self.id);
+        let transaction_handle = fire_unconfirmed_transactions(self.sender.as_mut().unwrap(), self.id);
+
+        self.handles.write().push(solution_handle);
+        self.handles.write().push(transaction_handle);
+    }
+
+    pub fn log_connections(&mut self) {
+        let self_clone = self.clone();
+        self.handles.write().push(tokio::task::spawn(async move {
+            loop {
+                let connections = self_clone.gateway().connected_peers().read().clone();
+                info!("{} connections", connections.len());
+                for connection in connections {
+                    debug!("  {}", connection);
+                }
+                sleep(Duration::from_secs(10)).await;
+            }
+        }));
+    }
+}
+
+impl TestNetwork {
+    // Creates a new test network with the given configuration.
+    pub fn new(config: TestNetworkConfig) -> Self {
+        if let Some(log_level) = config.log_level {
+            initialize_logger(log_level);
+        }
+
+        let (accounts, committee) = new_test_committee(config.num_nodes);
+
+        let mut primaries = HashMap::with_capacity(config.num_nodes as usize);
+        for (id, account) in accounts.into_iter().enumerate() {
+            let storage = Storage::new(committee.clone(), MAX_GC_ROUNDS);
+            let ledger = Box::new(MockLedgerService::new());
+            let primary = Primary::<CurrentNetwork>::new(account, storage, ledger, None, Some(id as u16)).unwrap();
+
+            let test_primary = TestPrimary { id: id as u16, primary, sender: None, handles: Default::default() };
+            primaries.insert(id as u16, test_primary);
+        }
+
+        Self { config, primaries }
+    }
+
+    // Starts each node in the network.
+    pub async fn start(&mut self) {
+        for primary in self.primaries.values_mut() {
+            // Setup the channels and start the primary.
+            let (sender, receiver) = init_primary_channels();
+            primary.sender = Some(sender.clone());
+            primary.run(sender.clone(), receiver, None).await.unwrap();
+
+            if self.config.fire_cannons {
+                primary.fire_cannons();
+            }
+
+            if self.config.log_connections {
+                primary.log_connections();
+            }
+        }
+
+        if self.config.connect_all {
+            self.connect_all().await;
+        }
+    }
+
+    // Starts the solution and trasnaction cannons for node.
+    pub fn fire_cannons(&mut self, id: u16) {
+        self.primaries.get_mut(&id).unwrap().fire_cannons();
+    }
+
+    // Connects a node to another node.
+    pub async fn connect(&self, id: u16, peer_id: u16) {
```

Comment:
```suggestion
    pub async fn connect_primaries(&self, first_id: u16, second_id: u16) {
```

Reply:
this API is a bit confusing; I was first expecting a broad network-related object to connect 2 nodes, but it seems that this treats `self` as belonging to a node already... but then its ID still needs to be provided?